pkg/cloud/services/ec2/launchtemplate.go

```diff
@@ -0,0 +1,329 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ec2
+
+import (
+	"encoding/base64"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/pkg/errors"
+	"k8s.io/utils/pointer"
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
+	expinfrav1 "sigs.k8s.io/cluster-api-provider-aws/exp/api/v1alpha3"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/awserrors"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/scope"
+)
+
+// GetLaunchTemplate returns the existing LaunchTemplate or nothing if it doesn't exist.
+// For now by name until we need the input to be something different
+func (s *Service) GetLaunchTemplate(id string) (*expinfrav1.AWSLaunchTemplate, error) {
+	if id == "" {
+		return nil, nil
+	}
+
+	s.scope.V(2).Info("Looking for existing LaunchTemplates")
+
+	input := &ec2.DescribeLaunchTemplateVersionsInput{
+		LaunchTemplateId: aws.String(id),
+		Versions:         aws.StringSlice([]string{expinfrav1.LaunchTemplateLatestVersion}),
+	}
+
+	out, err := s.EC2Client.DescribeLaunchTemplateVersions(input)
+	switch {
+	case awserrors.IsNotFound(err):
+		return nil, nil
+	case err != nil:
+		s.scope.Info("", "aerr", err.Error())
+	}
+
+	if len(out.LaunchTemplateVersions) == 0 {
+		return nil, nil
+	}
+
+	return s.SDKToLaunchTemplate(out.LaunchTemplateVersions[0])
+}
+
+// CreateLaunchTemplate generates a launch template to be used with the autoscaling group
+func (s *Service) CreateLaunchTemplate(scope *scope.MachinePoolScope, imageID *string, userData []byte) (string, error) {
+	s.scope.Info("Create a new launch template")
+
+	launchTemplateData, err := s.createLaunchTemplateData(scope, imageID, userData)
```

Comment:
Might be worth a follow up issue to figure out security implications of shared user data?

Reply:
good call. i opened an issue for this https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1872